rbd: added RBD features support for krbd

k0ste opened the pull request:
Added support for `object-map, fast-diff`

# Describe what this PR does #

This PR added support for all currently supported [krbd](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/block/rbd.c#n112) image features:

```
#define RBD_FEATURES_ALL	(RBD_FEATURE_LAYERING |		\
				 RBD_FEATURE_STRIPINGV2 |	\
				 RBD_FEATURE_EXCLUSIVE_LOCK |	\
				 RBD_FEATURE_OBJECT_MAP |	\
				 RBD_FEATURE_FAST_DIFF |	\
				 RBD_FEATURE_DEEP_FLATTEN |	\
				 RBD_FEATURE_DATA_POOL |	\
				 RBD_FEATURE_OPERATIONS)

/* Features supported by this (client software) implementation. */

#define RBD_FEATURES_SUPPORTED	(RBD_FEATURES_ALL)
```

The features is described in [man rbd](https://docs.ceph.com/en/latest/man/8/rbd/#cmdoption-rbd-image-feature)

```
Specifies which RBD format 2 feature should be enabled when creating an image. Multiple features can be enabled by repeating this option multiple times. The following features are supported:

· layering: layering support
· striping: striping v2 support
· exclusive-lock: exclusive locking support
· object-map: object map support (requires exclusive-lock)
· fast-diff: fast diff calculations (requires object-map)
· deep-flatten: snapshot flatten support
· journaling: journaled IO support (requires exclusive-lock)
· data-pool: erasure coded pool support
```


## Related issues ##

Mention any github issues relevant to this PR. Adding below line
will help to auto close the issue once the PR is merged.

Fixes: https://github.com/ceph/ceph-csi/issues/2513

k0ste committed:
rbd: added RBD features support for krbd

Added support for `object-map, fast-diff`

Signed-off-by: Konstantin Shalygin <k0ste@k0ste.ru>

pkalever commented on examples/rbd/storageclass.yaml:
```diff
@@ -33,10 +33,11 @@ parameters:
+   # `object-map`, `fast-diff` features. If `journaling` is enabled, must
+   # enable `exclusive-lock` too.
+   # imageFeatures: layering,journaling,exclusive-lock,object-map,fast-diff
+   imageFeatures: "layering"
```
We might want to do changes in the helm charts too?

Madhu-1 commented on internal/rbd/rbd_util_test.go:
```diff
@@ -86,33 +86,41 @@ func TestValidateImageFeatures(t *testing.T) {
 			"layering,exclusive-loc,journaling",
 			&rbdVolume{
-				Mounter: rbdNbdMounter,
+				Mounter: rbdDefaultMounter,
```
why this change is made? this is a negative case mounter should not matter.

Madhu-1 commented on internal/rbd/rbd_util_test.go:
```diff
@@ -86,33 +86,41 @@ func TestValidateImageFeatures(t *testing.T) {
 			"feature journaling requires exclusive-lock to be set",
 		},
 		{
-			"layering,exclusive-lock,journaling",
```
What is the reason for modifying the existing tests not adding new ones?

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -174,7 +174,18 @@ var supportedFeatures = map[string]imageFeature{
 	},
 	librbd.FeatureNameExclusiveLock: {
-		needRbdNbd: true,
+		needRbdNbd: false,
```
reason for removing nbd here?

k0ste replied:
The reason is:

```
  Warning  ProvisioningFailed    8s (x8 over 70s)   rbd.csi.ceph.com_ceph-csi-rbd-provisioner-6b9f54d9df-4slxv_a61dd25a-888b-4104-8e9b-e5ec09125b38  failed to provision volume with StorageClass "csi-rbd-sc": rpc error: code = InvalidArgument desc = feature exclusive-lock requires rbd-nbd for mount
```

In ancient times this feature was supported only by librbd, but currently support with krbd

pkalever reviewed (changes requested):
It will be great if we can add e2e tests for each of these features with krbd and rbd-nbd mounters?

Madhu-1 reviewed (changes requested):
@k0ste Thanks for the contribution please take care of the below items

* Update the existing documentations
* Update storageclass in  helm charts
* Add E2E for CreateVolume, CreateSnapshot, PVC-PVC Clone, PVC restore, etc for new image features  added
*

k0ste commented:
> * Update the existing documentations

Updated
> * Update storageclass in  helm charts

Updated
> * Add E2E for CreateVolume, CreateSnapshot, PVC-PVC Clone, PVC restore, etc for new image features  added

Test added

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -1051,6 +1051,40 @@ var _ = Describe("RBD", func() {
+					}
+					err = createRBDStorageClass(f.ClientSet, f, defaultSCName, nil, nil, deletePolicy)
+					if err != nil {
+						e2elog.Failf("failed to create storageclass with error %v", err)
```
Please add snapshot, create PVC from the snapshot, and PVC-PVC clone test also.

Madhu-1 reviewed (commented):
@k0ste CI is failing can you please check

tomiles commented:
@k0ste Thanks for this PR, looking forward to this merging. We've recently hit this issue too of using the features already  supported in the kernel but not being allowed by ceph-csi.

pkalever commented:
Honestly! I'm not quite convinced with the approach that we are taking in using various features of krbd. The feature set will be dependent on the kernel version (rbd driver loaded).

We should check for krbd supported features first and if the feature is available use krbd else we should fall back to rbd-nbd.

tomiles commented:
I agree that detecting it would be even better. But without this PR we are just blocking features supported in most of the current LTS distros kernel versions which also isn’t a good approach.

k0ste commented:
> We should check for krbd supported features first and if the feature is available use krbd else we should fall back to rbd-nbd

I think better is fail, because krbd & rbd-nbd performance is dramatically different. Admin should get error and make a self decision - use nbd or update kernel

pkalever commented:
> I think better is fail, because krbd & rbd-nbd performance is dramatically different. Admin should get error and make a self decision - use nbd or update kernel

Do you have any benchmarking results comparing both?
Right, architecture wise it makes sense to agree that the nbd performance would be a bit low, but I don't think they should be too bad or incomparable with krbd.

pkalever commented:
FYI: for detecting the runtime krbd features supported, here is the PR: https://github.com/ceph/ceph-csi/pull/2556

humblec commented:
> > We should check for krbd supported features first and if the feature is available use krbd else we should fall back to rbd-nbd
> 
> I think better is fail, because krbd & rbd-nbd performance is dramatically different. Admin should get error and make a self decision - use nbd or update kernel

@k0ste agreed, however if admin is flexible to fallback, may be the SC option ( `fallbackNbd`)  as in #2556 helps? just checking what is the normal practice/desire from admins  in different setups in these kind of scenarios .

nixpanic commented on e2e/rbd.go:
```diff
@@ -1479,6 +1514,7 @@ var _ = Describe("RBD", func() {
 					f)
 			})
 
+
```
no need to add the extra line here

nixpanic commented on e2e/rbd.go:
```diff
@@ -1515,6 +1551,43 @@ var _ = Describe("RBD", func() {
+			By("create a PVC-PVC clone with object-map,fast-diff image-features and bind it to an app", func() {
+				// pvc clone is only supported from v1.16+
+				if !k8sVersionGreaterEquals(f.ClientSet, 1, 16) {
+					Skip("pvc clone is only supported from v1.16+")
```
it seems that these `Skip()` calls cause skipping of all the following `By()` tests as well. Instead log a warning and return from the `By()` function.

Ceph-CSI does not try to support Kubernetes < 1.16 anymore, so this check can be remove completely.

nixpanic commented on e2e/rbd.go:
```diff
@@ -1466,6 +1466,41 @@ var _ = Describe("RBD", func() {
+				func() {
+					err := deleteResource(rbdExamplePath + "storageclass.yaml")
+					if err != nil {
+						e2elog.Failf("failed to delete storageclass with error %v", err)
```
nit: Please drop the `with error` wording, we're trying to remove the redundancy of reporting an error and additionally mentioning `with error`. I understand many existing tests still include it, so it would be nice if you can change it for your tests in an update of this PR.

nixpanic commented on e2e/rbd.go:
```diff
@@ -1720,6 +1793,7 @@ var _ = Describe("RBD", func() {
 				}
 			})
 
+
```
Please check for other additional unneeded empty lines.

nixpanic reviewed (commented):
Looks reasonable to me in general. Just a few nits mentioned related to the e2e tests.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -206,7 +206,18 @@ var (
+			needRbdNbd: false,
+			dependsOn:  []string{librbd.FeatureNameObjectMap},
+		},
+		librbd.FeatureNameDeepFlatten: {
```
I don't see e2e for deepflatten feature can you please include this in e2e also?

Madhu-1 replied:
@k0ste any specific reason for adding deep-flatten image feature? currently deep flatten is handled internally by cephcsi to make clones and snapshots independent of each other

k0ste replied:
Deep-flatten may be needed by external administration features, out of Kubernetes scope

Madhu-1 replied:
as I mentioned earlier deep flattening is already handled by cephcsi, I think we can have a separate PR with the use-case that it will solve.

humblec commented on e2e/rbd.go:
```diff
@@ -1515,6 +1551,43 @@ var _ = Describe("RBD", func() {
+					defaultSCName,
+					nil,
+					map[string]string{
+						"imageFeatures": "layering,exclusive-lock,object-map,fast-diff",
```
may be add `deep-flatten` too ?

humblec commented on e2e/rbd.go:
```diff
@@ -1515,6 +1551,43 @@ var _ = Describe("RBD", func() {
+
+			By("create a PVC-PVC clone with object-map,fast-diff image-features and bind it to an app", func() {
+				// pvc clone is only supported from v1.16+
+				if !k8sVersionGreaterEquals(f.ClientSet, 1, 16) {
```
This is no longer needed, please see #2617 , in short, the version check below 1.20 is not needed any more, so you can remove this.

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@k0ste do you need any help in moving this forward? if you are busy I can cherry-pick your commit and take this to completion by this week. please let me know what do you think?

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -206,7 +206,18 @@ var (
 		},
 		librbd.FeatureNameExclusiveLock: {
-			needRbdNbd: true,
+			needRbdNbd: false,
```
as there is no mandatory requirement for the nbd mounted here, unit tests also need to be corrected for this one.

k0ste replied:
What you mean? `needRbdNbd` is already `false`

Madhu-1 replied:
```=== RUN   TestGetCephClientLogFileName/test_for_empty_id
=== CONT  TestValidateImageFeatures
    rbd_util_test.go:176: 
        	Error Trace:	rbd_util_test.go:176
        	Error:      	Error message not equal:
        	            	expected: "feature exclusive-lock requires rbd-nbd for mounter"
        	            	actual  : "feature journaling requires rbd-nbd for mounter"
        	Test:       	TestValidateImageFeatures
    rbd_util_test.go:176: 
        	Error Trace:	rbd_util_test.go:176
        	Error:      	Error message not equal:
        	            	expected: "feature exclusive-lock requires rbd-nbd for mounter"
        	            	actual  : "feature journaling requires rbd-nbd for mounter"
        	Test:       	TestValidateImageFeatures
--- FAIL: TestValidateImageFeatures (0.00s)
```
please check the [go test](https://github.com/ceph/ceph-csi/runs/4427733123?check_suite_focus=true#step:3:1544) failure you will get what needs to be changed

Madhu-1 reviewed (changes requested):
need to adjust the go-tests and fix go linter errors

Madhu-1 commented:
I have tested this on my local minikube machine,  as expected the mapping fails as minikube uses a lower kernel version (4.9.x).


```text
rbd info replicapool/csi-vol-c3f1eb8a-5650-11ec-af4c-ca30c5bfe2cb
rbd image 'csi-vol-c3f1eb8a-5650-11ec-af4c-ca30c5bfe2cb':
	size 1 GiB in 256 objects
	order 22 (4 MiB objects)
	snapshot_count: 0
	id: 11cbfe59d66d
	block_name_prefix: rbd_data.11cbfe59d66d
	format: 2
	features: layering, exclusive-lock, object-map, fast-diff
	op_features: 
	flags: 
	create_timestamp: Mon Dec  6 04:55:42 2021
	access_timestamp: Mon Dec  6 04:55:42 2021
	modify_timestamp: Mon Dec  6 04:55:42 2021

```


* Pod describe ouput
```text
 Warning  FailedMount             0s (x4 over 4s)  kubelet                  MountVolume.MountDevice failed for volume "pvc-47d81c8c-76de-420e-8c63-a130641dbec4" : rpc error: code = Internal desc = rbd: map failed with error an error (exit status 6) occurred while running rbd args: [--id csi-rbd-node -m 10.104.3.155:6789 --keyfile=***stripped*** map replicapool/csi-vol-c3f1eb8a-5650-11ec-af4c-ca30c5bfe2cb --device-type krbd --options noudev], rbd error output: rbd: sysfs write failed
rbd: map failed: (6) No such device or address
```

* dmesg logs from the minikube node

```log
$ uname -r
4.19.202

$ dmesg
...
[  892.675164] libceph: mon0 10.104.3.155:6789 session established
[  892.675246] libceph: mon0 10.104.3.155:6789 socket closed (con state OPEN)
[  892.675262] libceph: mon0 10.104.3.155:6789 session lost, hunting for new mon
[  892.675594] libceph: mon0 10.104.3.155:6789 session established
[  892.675693] libceph: client4642 fsid e65ffd0f-8425-443f-97e8-276410d951cb
[  892.683899] rbd: image csi-vol-c3f1eb8a-5650-11ec-af4c-ca30c5bfe2cb: image uses unsupported features: 0x18
[  924.899178] libceph: mon0 10.104.3.155:6789 session established
[  924.899249] libceph: mon0 10.104.3.155:6789 socket closed (con state OPEN)
[  924.899265] libceph: mon0 10.104.3.155:6789 session lost, hunting for new mon
[  924.899601] libceph: mon0 10.104.3.155:6789 session established
[  924.899705] libceph: client4657 fsid e65ffd0f-8425-443f-97e8-276410d951cb
[  924.905023] rbd: image csi-vol-c3f1eb8a-5650-11ec-af4c-ca30c5bfe2cb: image uses unsupported features: 0x18
$ 
```

 @k0ste if you can remove the e2e  changes we can continue on reviews other changes

k0ste commented:
> @k0ste if you can remove the e2e changes we can continue on reviews other changes

What exactly e2e changes? I was included everything tips that your team left🤔

Madhu-1 commented:
> > @k0ste if you can remove the e2e changes we can continue on reviews other changes
> 
> What exactly e2e changes? I was included everything tips that your team leftthinking

the newly added e2e test will fail in our e2e environment as we are using a minikube to create a kubernetes cluster. please check https://github.com/ceph/ceph-csi/pull/2514#issuecomment-986443038

k0ste commented:
`deep-flatten` was removed as requested

Madhu-1 commented:
Please run `CONTAINER_CMD=docker make containerized-test TARGET=go-test` and `run: CONTAINER_CMD=docker make containerized-test TARGET=go-lint` locally to see what is the issue if you see any go-test or go-lint test failure.

k0ste commented:
Some CI test failed due:
```
kernel 4.19.202 does not support required features
```

Madhu-1 commented:
> Some CI test failed due:
> 
> ```
> kernel 4.19.202 does not support required features
> ```

CI is not going to pass as minikube comes with a kernel version of `4.19.202` which does not support fast-diff,obj-map... image feature. i tested this one and mentioned the same at https://github.com/ceph/ceph-csi/pull/2514#issuecomment-986443038 and also you need to remove the e2e changes you are adding as I mentioned here https://github.com/ceph/ceph-csi/pull/2514#issuecomment-986545025.

k0ste commented:
> > Some CI test failed due:
> > ```
> > kernel 4.19.202 does not support required features
> > ```
> 
> CI is not going to pass as minikube comes with a kernel version of `4.19.202` which does not support fast-diff,obj-map... image feature. i tested this one and mentioned the same at [#2514 (comment)](https://github.com/ceph/ceph-csi/pull/2514#issuecomment-986443038) and also you need to remove the e2e changes you are adding as I mentioned here [#2514 (comment)](https://github.com/ceph/ceph-csi/pull/2514#issuecomment-986545025).

e2e changes was removed

pkalever commented on internal/rbd/rbd_util_test.go:
```diff
@@ -88,29 +88,53 @@ func TestValidateImageFeatures(t *testing.T) {
 			"layering,journaling",
 			&rbdVolume{
-				Mounter: rbdNbdMounter,
+				Mounter: rbdDefaultMounter,
```
Why do we need this change?

Madhu-1 replied:
no harm in this one.

pkalever reviewed (changes requested):
IMO, we shouldn't have removed e2e tests rather defended the test with `util.CheckKernelSupport` for the minimum required version, that way the CI will not bailout. 

@k0ste can this be reverted with added defence check?

cc: @Madhu-1

Madhu-1 commented:
> IMO, we shouldn't have removed e2e tests rather defended the test with `util.CheckKernelSupport` for the minimum required version, that way the CI will not bailout.

Yes, but I don't think it's useful as this will never get executed in near time as we are still on 4.9 and what we need is 5.3 and moreover, we need to rework that one to check the kernel version on all nodes and schedule pod for that node.

Madhu-1 reviewed (approved):
LGTM

Madhu-1 commented:
@mergifyio rebase

pkalever commented:
> Yes, but I don't think it's useful as this will never get executed in near time as we are still on 4.9 and what we need is 5.3 and moreover, we need to rework that one to check the kernel version on all nodes and schedule pod for that node.

Soon minikube will have 5.x, minkube community is working on it. I just wanted to avoid rework on it later.

@k0ste I don't insist on having the e2e, but please create an issue for this so that we don't miss the e2e coverage later.

pkalever reviewed (approved):
Well, the changes LGTM. My only concern is on the e2e tests. Assuming @k0ste will open an issue for e2e tests later. Approving this now.
